EventFilter/DTRawToDigi/plugins/DTuROSRawToDigi.cc

```diff
@@ -0,0 +1,536 @@
+
+	    DTDigi digi(wire, tdcTime, hitOrder[channelIndex.getCode()]);
+	    digis.insertDigi(detId.layerId(),digi);
+
+	  }
+
+	}
+        else if ( selector == 3 ) { // error word
+
+	  if (  debug_ ) edm::LogWarning("dturos_unpacker") << "Error word [" << std::dec << k << "] : " << std::hex << dataWord 
+							    << std::dec << " in slot " << slot << " in crate " << crate;
+
+	  int error  = ( dataWord >> 32 ) & 0x1FFFFFFF; // positions  32 -> 60
+	  rwords.seterror(error);
+
+	}
+
+
+	if ((dataWord & 0x1FFFFFFF) == 0x1FFFFFFF) continue;
+
+	if ( selector2 == 0 ) { // TDC word
+
+	  int tdcTime    = ( dataWord ) & 0x3FFF;       // positions   0 -> 13
+	  int tdcChannel = ( dataWord >> 14 ) & 0x1F;   // positions  14 -> 18
+	  int tdcId      = ( dataWord >> 19 ) & 0x3;    // positions  19 -> 20
+	  int link       = ( dataWord >> 21 ) & 0x7F;   // positions  21 -> 27
+
+
+	  if (tdcTime == 16383) continue;
+
+
+	  int dduId = theDDU(crate, slot, link);
+	  int rosId = theROS(crate, slot, link);
+	  int robId = theROB(crate, slot, link);
+
+
+	  DTROChainCoding channelIndex(dduId, rosId, robId, tdcId, tdcChannel);
+	  if (hitOrder.find(channelIndex.getCode()) == hitOrder.end()) hitOrder[channelIndex.getCode()] = 0;
+	  else hitOrder[channelIndex.getCode()]++;
+
+
+	  int wheelId, stationId, sectorId, slId,layerId, cellId;
+	  if ( ! mapping->readOutToGeometry(dduId, rosId, robId, tdcId, tdcChannel,
+					    wheelId, stationId, sectorId, slId, layerId, cellId)) {
+
+	    DTWireId detId = DTWireId(wheelId, stationId, sectorId, slId, layerId, cellId);
+	    int wire = detId.wire();
+
+	    DTDigi digi(wire, tdcTime, hitOrder[channelIndex.getCode()]);
+	    digis.insertDigi(detId.layerId(),digi);
+
+	  }
+
+	}
+        else if ( selector2 == 1 ) { // error word
+
+	  if (  debug_ ) edm::LogWarning("dturos_unpacker") << "Error word [" << std::dec << k << "] : " << std::hex << dataWord 
+							    << std::dec << " in slot " << slot << " in crate " << crate;
+
+	  int error  = ( dataWord ) & 0x1FFFFFFF;       // positions   0 -> 28
+	  rwords.seterror(error);
+
+	}
+
+      }
+
+    }
+
+
+    ++DTuROSiterator;
+    dataWord  = (*DTuROSiterator); // Trailer AMC
+
+
+    rwords.settrailer(dataWord);
+    fwords.setuROS(slot, rwords);
+
+  }  // end for-loop container content
+
+
+  return;
+}
+
+
+int DTuROSRawToDigi::theDDU(int crate, int slot, int link) {
+
+  if (crate == 1368) {
+    if (slot < 7) return 770;
+    return 771;
+  }
+
+  if (crate == 1370) {
+    if (slot > 6) return 773;
+    return 774;
+  }
+
+  return 772;
+}
+
+
+int DTuROSRawToDigi::theROS(int crate, int slot, int link) {
+
+  if (slot%6 == 5) return link+1;
+
+  int ros = (link/24) + 3*(slot%6) - 2;
+  return ros;
+}
+
+
+int DTuROSRawToDigi::theROB(int crate, int slot, int link) {
+
+  if (slot%6 == 5) return 23;
+ 
+  int rob = link%24;
+  if (rob < 15) return rob;
+  if (rob == 15) return 24;
+  return rob-1;
+}
+
+
+void DTuROSRawToDigi::calcCRC(long word, int& myC) {
```

Comment:
If I am not wrong, this function looks identical to 
https://cmssdt.cern.ch/lxr/source/EventFilter/L1TXRawToDigi/plugins/L1TTwinMuxRawToDigi.cc#0934
If it has to be so by construction, you could avoid code duplication (and all risks of possible mistakes and bad maintenace connected to it) by defining one single helper function that will get called by all them.

Reply:
Dear Andrea
We are installing a new set of boards this shutdown and it is mandatory to be able to decode the data we extract. We don´t want to touch nor validate the TwinMux or DTTF unpackers. So we don´t want to make any change there. 
We don´t know how this single helper function can be created and I really hope this does not become a showstopper for the DT Phase 1 upgrade.
Is there any way we can proceed?
Thanks
Cristina (DT Project Manager)

Reply:
Hi Cristina.
I don't think it can be a big issue defining this function in some common location (EventFilter/Utilities? or maybe even here, in view of going toward single packer/unpacker modules as it was requested in the former comments to this PR).
However, if it is going to become a showstopper or the DT Phase 1 upgrade feel free to descope it to another, less urgent, PR, and concentrate here on the other issues raised (say the  inclusion in the 2018 workflows)

Reply:
@cristosaurio : I think I misunderstood your question as "we need to merge this quickly, we will fix the code later".  If you plan to need this code inside 10_0_X, then there is plenty of time to adjust it, without the need to wait for some later PR. (If you need it earlier on, you should probably foresee some backport, instead).

The fix/cleaning we ask for could be as simple as follows. just define that calcCRC method in one class, and link and reuse the same in the other classes that need the very same method. That class could be one of the existing ones, or the methods can be put in a separate utility piece of code which only includes that function(s), and that can get linked and used by all classes in CMSSW that need it.

If you or the main developers of his PR want to discuss it, you are welcome to join the reco meeting this Friday, were we can iterate it.

Reply:
Hi Andrea My problem is to assign the scarce manpower resources we have on this (which seems just cosmetics because the code works perfectly fine) with respect to the rest of critical things we need to get done in DT. If you really think this is mandatory and not just cosmetics, then I will drop something else and get someone to do this. Thanks